rfc/20241023-rfc-tracking-issues.md

```diff
@@ -0,0 +1,209 @@
+
+Once an RFC's tracking issue has been closed to represent that work on it is complete, the RFC content is frozen as a historical record of that work. If we later plan other work in the same area -- which might in some cases conflict with what was previously proposed and accepted -- we should write a new RFC that describes its proposed changes relative to the outcome of the previous RFC.
+
+One exception to the frozen state: if a new RFC is proposed that materially changes a decision made and recorded in an earlier accepted RFC, the PR for the new RFC should add a note in each relevant part of the older RFC that mentions that the decision has been changed and links to the new RFC, so that it's clear to a future reader of the old RFC that part of it has been obsoleted retroactively. Consider using [GitHub Markdown's "alert" extensions](https://github.blog/changelog/2023-12-14-new-markdown-extension-alerts-provide-distinctive-styling-for-significant-content/) to clearly separate the later-added content from the original frozen content, and add the new markers in the same PR that introduces the new RFC so that the changes are more strongly connected in the history.
+
+The "chains" of linked RFCs that might be caused by the frozen state could potentially become quite long over time for product areas that experience frequent change, but thankfully older RFCs tend to become less and less relevant over time. Any significant feature of the product or its codebases should also be documented in some separate location in a "how it currently is" shape, rather than a "how we changed it over time" shape, so that it isn't necessary to piece together information on currently-available features by trawling through the RFC history unless the historical change decisions _themselves_ are the point of interest.
+
+Some existing locations for longer-lived supporting documentation include:
+
+- [The OpenTofu docs](https://opentofu.org/docs/), maintained in the `website` directory in the root of the main OpenTofu repository, for end-user oriented documentation.
+- The `docs` directory in the root of the main OpenTofu repository, for developer-oriented descriptions of system architecture and the design of non-trivial features.
+- The `CONTRIBUTING.md` file and the RFC `README.md` file in the OpenTofu repository, for contributor oriented documentation about the development process rather than the code itself.
+- The documentation comments directly within our Go packages, attached to individual symbols and whole packages, for developer-oriented descriptions of fine-grain details specific the implementation of certain features.
```

Comment:
Yeah, the decision between package-level Go doc commands vs. a separate `README.md` in the directory has always been a little unclear to me. I don't have a strong opinion either way and have done both at different times. 

I guess I've typically preferred to use Go doc comments and used a separate readme only when the docs got particularly long or benefitted from Markdown formatting, but it is true that the Go doc comments are not so easy to find for someone who doesn't already know how to construct a `pkg.go.dev` URL for one of our internal packages, whereas the README.md is rendered prominently in the GitHub UI.

Again, I think this would be a good topic for a separate RFC focused on documentation. As with the others, I included this only as an example of what we might currently do. I will update this to mention that a `README.md` in the package directory is a valid alternative though, since that _is_ describing something we already do rather than a new proposal.